DDD-1.md

```diff
@@ -0,0 +1,173 @@
+
+### Recover database schema from multiple partitions in one pass ([#14](https://github.com/sugarcrm/debezium/issues/14))
+
+Currently the internal schema representation used by the connector is capable of representing multiple databases (e.g. for the MySQL connector) but isn't capable of initializing from the database history of multiple partitions. It should be possible to initialize it multiple times, once for each partition but it may be suboptimal, especially if the database history topic and the number of partitions are large.
+
+Instead, the schema could recover from multiple offsets in one pass.
+
+**Before**:
+
+```java
+for (message : messages) {
+  if (message.partition == partition && message.offset < offset) {
+    apply(message);
+  }
+}
+```
+
+**After**:
+
+```java
+for (message : messages) {
+  if (offsets.contains(message.partition)
+      && message.offset < offsets.get(message.partition)) {
+      apply(message);
+    }
+  }
+}
+```
+
+### Rework the "real" database name detection logic ([#21](https://github.com/sugarcrm/debezium/issues/21))
+
+Currently, the "real" name of the source database is detected right after connecting to the database at the connection level. In order to make the connection reusable across multiple databases (https://github.com/sugarcrm/debezium/issues/10), we need to:
+
+1. Detect the "real" name of all databases the task is working with.
+2. Make the "real" name available to all components that use it right now.
+
+The most natural place for that seems to be the `SqlServerTaskPartition` and its `Provider` being introduced as part of #15.
+
+### Use fully-qualified names for all tables and procedures in Debezium connector for SQL Server ([#10](https://github.com/sugarcrm/debezium/issues/10))
+
+See details in [#10](https://github.com/sugarcrm/debezium/issues/10). This is where the connector-level partitions defined in the beginning come into play. Given a partition, the connector will be able to call `partition.getDatabase()` and perform a query / execute a statement on a connection that isn't scoped to any specific database.
+
+### Introduce `Partitioned*` APIs ([#13](https://github.com/sugarcrm/debezium/pull/13), [#16](https://github.com/sugarcrm/debezium/issues/16), [#18](https://github.com/sugarcrm/debezium/issues/18))
+
+####  `Partitioned*` interfaces
+
+Despite the fact that there are multiple connectors that can benefit from being aware, some connectors are nonpartitioned by their nature. For instance, the MySQL connector consumes all the changes from a nonpartitioned binlog, and it would make sense to have its API and the implementation affected by the changes being proposed as little as possible.
+
+In order to accomodate partition-aware and -unaware components under the same framework, it's proposed to introduce the `Partitioned` analogues of the following interfaces:
+
+1. `SnapshotChangeEventSource` → `PartitionedSnapshotChangeEventSource`
```

Comment:
Why is it that the change event source needs to have awareness of partitioning? IIRC, during our call we assuming that the existing implementation could effectively work on multiple databases, only the configuration which the task receives needs to be adjusted?

Reply:
> IIRC, during our call we assuming that the existing implementation could effectively work on multiple databases, only the configuration which the task receives needs to be adjusted?

Yes, we did assume that. I stumbled on a few challenges during the implementation. Here's the reasoning:

1. Some connectors like the one for MySQL are nonpartitioned by their nature. Making it implement the partitioned API looks suboptimal. Both the snapshot and the streaming sources of a partition-aware connector should return to the coordinator as soon as there are no more changes in order to allow another partition to be processed. But for MySQL, there are no other partitions by definition.
2. Having two sets of APIs will allow for easier migration. First, a connector implements the old API. Second, the connector implements both the new and the old APIs with the ability to opt-in and opt-out. The connector stops implementing the old API.

Reply:
>Both the snapshot and the streaming sources of a partition-aware connector should return to the coordinator as soon as there are no more changes in order to allow another partition to be processed.

I don't understand that part yet. Yes, the snapshot source should return once it has processed all partions, be it one, or multiple. But the streaming souce keeps running until the connector gets stopped. In my perception, the loop iterating through the databases would be _within_ the streaming source of a connector supporting multiple databases.

Reply:
@morozov, so I believe I understand now, that you envision to have one streaming change event source _per partition_ (i.e. database), correct? So if e.g. a task processes three partitions, there'd be three streaming change event sources, and the coordinator cycles through them. Is that what you have in mind? If so, I don't get yet the change of the pending PR, where offsets are moved to method parameters. Would this be actually necessary in this case? I feel I'm still misunderstanding something here.

Reply:
This is outdated. I removed this section.